scripts/publish.py

```diff
@@ -0,0 +1,123 @@
+
+# install dependencies:
+# pip3 install toposort
+
+import argparse
+import os
+import re
+import subprocess
+import time
+
+from typing import Optional
+from toposort import toposort_flatten
+
+
+# TODO find this automatically
+target_version = "1.0.0-alpha01.1"
+
+# TODO generate this by parsing toml files
+dep_graph = {
+    "wasmer-types": set([]),
+    "wasmer-vm": set(["wasmer-types"]),
+    "wasmer-compiler": set(["wasmer-vm", "wasmer-types"]),
+    "wasmer-object": set(["wasmer-types", "wasmer-compiler"]),
+    "wasmer-engine": set(["wasmer-types", "wasmer-vm", "wasmer-compiler"]),
+    "wasmer-compiler-singlepass": set(["wasmer-types", "wasmer-vm", "wasmer-compiler"]),
+    "wasmer-compiler-cranelift": set(["wasmer-types", "wasmer-vm", "wasmer-compiler"]),
+    "wasmer-compiler-llvm": set(["wasmer-types", "wasmer-vm", "wasmer-compiler"]),
+    "wasmer-engine-jit": set(["wasmer-types", "wasmer-vm", "wasmer-compiler", "wasmer-engine"]),
+    "wasmer-engine-native": set(["wasmer-types", "wasmer-vm", "wasmer-compiler", "wasmer-engine",
+                                 "wasmer-object"]),
+    "wasmer": set(["wasmer-vm", "wasmer-compiler-singlepass", "wasmer-compiler-cranelift",
+                   "wasmer-compiler-llvm", "wasmer-compiler", "wasmer-engine", "wasmer-engine-jit",
+                   "wasmer-engine-native", "wasmer-types"]),
+    "wasmer-cache": set(["wasmer"]),
+    "wasmer-wasi": set(["wasmer"]),
+    "wasmer-wasi-experimental-io-devices": set(["wasmer-wasi"]),
+    "wasmer-emscripten": set(["wasmer"]),
+}
+
+# where each crate is located in `lib`
+# this can also be generated from the toml files
+location = {
+    "wasmer-types": "wasmer-types",
+    "wasmer-vm": "vm",
+    "wasmer-compiler": "compiler",
+    "wasmer-object": "object",
+    "wasmer-engine": "engine",
+    "wasmer-compiler-singlepass": "compiler-singlepass",
+    "wasmer-compiler-cranelift": "compiler-cranelift",
+    "wasmer-compiler-llvm": "compiler-llvm",
+    "wasmer-engine": "engine",
+    "wasmer-engine-jit": "engine-jit",
+    "wasmer-engine-native": "engine-native",
+    "wasmer-cache": "cache",
+    "wasmer": "api",
+    "wasmer-wasi": "wasi",
+    "wasmer-emscripten": "emscripten",
+    "wasmer-wasi-experimental-io-devices": "wasi-experimental-io-devices",
+}
+
+dry_run = False
+
+def get_latest_version_for_crate(crate_name: str) -> Optional[str]:
+    output = subprocess.run(["cargo", "search", crate_name], capture_output=True)
+    rexp_src = '^{} = "([^"]+)"'.format(crate_name)
+    prog = re.compile(rexp_src)
+    haystack = output.stdout.decode("utf-8")
+    for line in haystack.splitlines():
+        result = prog.match(line)
+        if result:
+            return result.group(1)
+
+def is_crate_already_published(crate_name: str) -> bool:
+    found_string = get_latest_version_for_crate(crate_name)
+    if found_string is None:
+        return False
+
+    return target_version == found_string
```

Comment:
Strictly speaking this only implements "is crate already published" as long as there is no newer published version of the crate than we're trying to publish. Which shouldn't happen if all releases use this script and set the version correctly, but it would be nice if it could detect or recover from manual crate releases. (Again, this is not a situation we intend to put ourselves in, but unlike the symlink case, I can see someone tripping over this for real, probably during a fire drill related to why we had a manual release of one crate to begin with.)

Reply:
Yeah, I'm not too worried about it  -- I think it's good enough for now; if this fails and we republish something, then it will just cause `cargo publish` to fail, which will either cause python to throw an exception or continue executing as normal, either outcome seems fine

Reply:
Also having to understand versions seems a bit too advanced for this script -- the script expects all the Cargo.toml files to have the correct version to publish already... the hardcoded version should really just come from the `toml` file itself to allow the script to work correctly with transient network failures or whatever

Reply:
It's just the mismatch between what it does and the name. If it were "is_version_already_published" I'd be fine with it since it does exactly what it says on the tin, but this kind of minor semantic mismatch is exactly the kind of thing I'm looking for in a code review.

Reply:
oh, you are right that if a later version is published, that won't show up with `cargo search`, but I don't think there's any way to check if the crate is already published given a later version without attempting to download it, curling crates.io, or attempting to publish it.  I think it's good enough for now as it will serve its purpose of speeding up retries in case of failure and should continue working even if this check fails.